include/ofi_hmem.h

```diff
@@ -105,17 +101,17 @@ hsa_status_t ofi_hsa_amd_reg_dealloc_cb(void *ptr,
 					hsa_amd_deallocation_callback_t cb,
 					void *user_data);
 
-hsa_status_t ofi_hsa_amd_memory_lock(void *host_ptr, size_t size,
-				     hsa_agent_t *agents, int num_agents,
-				     void **agent_ptr);
-hsa_status_t ofi_hsa_amd_memory_unlock(void *host_ptr);
-
 #endif /* HAVE_ROCR */
 
 struct ofi_hmem_ops {
 	bool initialized;
 	int (*init)(void);
 	int (*cleanup)(void);
+	int (*async_copy_to_hmem)(uint64_t device, void *dest, const void *src,
+			    size_t size, void **stream);
+	int (*async_copy_from_hmem)(uint64_t device, void *dest, const void *src,
+			      size_t size, void **stream);
```

Comment:
Stream as output looks wrong.  I would have expected this to be input to indicate which stream to perform the copy, with some sort of signal or event being used to indicate the completion.  All the major GPUs support this type of operation, can we get an overview of how it would be done on each?

The application may want some control over which stream is used, versus the library picking it.

Reply:
HSA has its own signal infrastructure which is used to indicate when the operation is complete. A stream is an abstract object which is meant to track several set of iovec operations. For example if the iovec has several entries and you do an async copy on each. This intent of this API is to allow the ROCR specific code to manage the signals and return an abstract object, stream, which is then used by the caller to query when the operations are complete. I don't see why the application would want to control that. 

The concept of stream here does not map to the hip or cuda stream. In hip, if you want to use an API like hipMemcpyAsync(), you'll need  to create the stream using hipStreamCreate() then pass that to hipMemcpyAsync(). Then you destroy it once you're done. I suppose we can follow the same flow, where we add new APIs to create and destroy streams and then pass the stream to the async_copy API.  

However, this will complicate the API for a non-existent use case. As you mentioned in a  previous comment "Until we have another implementation, it's not worth carrying the extra complexity in the hope that it may someday be useful. It's too difficult to predict what might be needed."

Reply:
I am wondering whether the name 'stream' is causing confusion, isn't this really behaving more like a request in MPI, i.e. you can use that to check for completion, with the difference being that it contains more than one element potentially.